### PR TITLE
Implemented decode methods for the Geometry codecs

### PR DESCRIPTION
**Pull request by rozza**
JAVA-2247

### CODE REVIEW

File: driver-core/src/main/com/mongodb/client/model/geojson/codecs/NamedCoordinateReferenceSystemCodec.java [hunk 1]
```diff
@@ -47,6 +51,10 @@ public void encode(final BsonWriter writer, final NamedCoordinateReferenceSystem
 
     @Override
     public NamedCoordinateReferenceSystem decode(final BsonReader reader, final DecoderContext decoderContext) {
-        throw new UnsupportedOperationException("Not implemented yet!");
+        CoordinateReferenceSystem crs = decodeCoordinateReferenceSystem(reader);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/geojson/codecs/NamedCoordinateReferenceSystemCodec.java]: Nit: this method isn't tested, though the helper it calls is.

File: driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy [hunk 2]
```diff
@@ -59,11 +66,44 @@ class GeometryCollectionCodecSpecification extends Specification {
...
+    def 'should throw when decoding invalid documents'() {
+        when:
+        codec.decode(new BsonDocumentReader(parse(invalidJson)), DecoderContext.builder().build())
+
+        then:
+        thrown(CodecConfigurationException)
+
+        where:
+        invalidJson << [
+                '''{ type: "GeometryCollect",
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy]: Nit: for full coverage add one with no type field in each codec spec

Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy]: Done

File: driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy [hunk 3]
```diff
@@ -42,11 +45,15 @@ class GeometryCollectionCodecSpecification extends Specification {
         codec.encode(writer, geometryCollection, context)
 
         then:
-        writer.document == parse('{ type: "GeometryCollection",' +
-                                 '    geometries: [' +
-                                 '      { type: "Point", coordinates: [100.0, 0.0]},' +
-                                 '      { type: "LineString", coordinates: [ [101.0, 0.0], [102.0, 1.0] ]}' +
-                                 '    ]}')
+        writer.document == parse('''{ type: "GeometryCollection",
+                geometries: [{ type: "Point", coordinates: [100.0, 0.0]},
```
Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy]: Done

File: driver-core/src/main/com/mongodb/client/model/geojson/codecs/NamedCoordinateReferenceSystemCodec.java (hunk 1 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/geojson/codecs/NamedCoordinateReferenceSystemCodec.java]: Done

File: driver-core/src/main/com/mongodb/client/model/geojson/codecs/AbstractGeometryCodec.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+abstract class AbstractGeometryCodec<T extends Geometry> implements Codec<T> {
+    private final CodecRegistry registry;
+
+    AbstractGeometryCodec(final CodecRegistry registry) {
+        this.registry = registry;
+    }
+
+    @Override
+    public void encode(final BsonWriter writer, final Geometry value, final EncoderContext encoderContext) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/geojson/codecs/AbstractGeometryCodec.java]: value should be of type T, not type Geometry

File: driver-core/src/main/com/mongodb/client/model/geojson/codecs/AbstractGeometryCodec.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import org.bson.BsonWriter;
+import org.bson.codecs.Codec;
+import org.bson.codecs.DecoderContext;
+import org.bson.codecs.EncoderContext;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import static com.mongodb.client.model.geojson.codecs.GeometryDecoderHelper.decodeGeometry;
+import static com.mongodb.client.model.geojson.codecs.GeometryEncoderHelper.encodeGeometry;
+
+abstract class AbstractGeometryCodec<T extends Geometry> implements Codec<T> {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/geojson/codecs/AbstractGeometryCodec.java]: You can simplify a bit further by taking an instance of Class<T> in the constructor and implementing getEncoderClass method here as well.

File: driver-core/src/main/com/mongodb/client/model/geojson/codecs/GeometryDecoderHelper.java [hunk 6]
```diff
@@ -503,7 +448,11 @@ private static String decodeCoordinateReferenceSystemProperties(final BsonReader
     }
 
     private static void validateIsDocument(final BsonReader reader) {
-        if (reader.getCurrentBsonType() != BsonType.DOCUMENT) {
+        BsonType currentType = reader.getCurrentBsonType();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/geojson/codecs/GeometryDecoderHelper.java]: What bug did this fix?

File: driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy [hunk 7]
```diff
@@ -33,37 +41,146 @@ class GeometryCollectionCodecSpecification extends Specification {
...
+        writer.document == parse('{ type: "GeometryCollection", geometries: []}')
+
+        when:
+        def decodedGeometryCollection = codec.decode(new BsonDocumentReader(writer.document), DecoderContext.builder().build())
+
+        then:
+        geometryCollection == decodedGeometryCollection
+    }
+
     def 'should encode'() {
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy]: Missed this the first time... Consider renaming all the "should encode" methods to "should round trip" or "should encode and decode"

File: driver-core/src/main/com/mongodb/client/model/geojson/codecs/AbstractGeometryCodec.java (hunk 5 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/geojson/codecs/AbstractGeometryCodec.java]: Nice - Done

File: driver-core/src/main/com/mongodb/client/model/geojson/codecs/AbstractGeometryCodec.java (hunk 4 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/geojson/codecs/AbstractGeometryCodec.java]: Done

File: driver-core/src/main/com/mongodb/client/model/geojson/codecs/GeometryDecoderHelper.java (hunk 6 above)
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/geojson/codecs/GeometryDecoderHelper.java]: Not a bug per se, when decoding a CoordinateReferenceSystem we jump straight to this method and we haven't read anything.

File: driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy (hunk 7 above)
Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/client/model/geojson/codecs/GeometryCollectionCodecSpecification.groovy]: Done

**Pull request closed without merging**